### PR TITLE
[Feature]: Support Batch for Faster R-CNN

### PR DESCRIPTION
**Pull request by hhaAndroid**
[Feature]: Support Batch for Faster R-CNN

### CODE REVIEW

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 1]
```diff
@@ -277,33 +277,63 @@ def get_bboxes(self,
                    cfg=None):
         if isinstance(cls_score, list):
             cls_score = sum(cls_score) / float(len(cls_score))
-        scores = F.softmax(cls_score, dim=1) if cls_score is not None else None
+
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:280]: add unitest for batch inference， especially for empty rois with shape (bs,0,5).

add the docstr of this file，expecially for `rois`

File: mmdet/models/roi_heads/test_mixins.py [hunk 2]
```diff
@@ -56,50 +56,93 @@ def simple_test_bboxes(self,
...
+            assert len(
+                img_metas
+            ) == 1, 'Only support one input image while in exporting to ONNX'
+            img_shapes = img_metas[0]['img_shape_for_onnx']
         else:
             img_shapes = tuple(meta['img_shape'] for meta in img_metas)
         scale_factors = tuple(meta['scale_factor'] for meta in img_metas)
 
-        # split batch bbox prediction back to each image
+        if isinstance(proposals, list):
```
Comment (jshilong) [inline mmdet/models/roi_heads/test_mixins.py:69]: explain why we add padding to form a batch dimension

File: mmdet/models/roi_heads/test_mixins.py [hunk 3]
```diff
@@ -56,50 +56,93 @@ def simple_test_bboxes(self,
...
+            rois = proposals
+
+        batch_index = torch.arange(
+            rois.size(0), device=rois.device).float().view(-1, 1, 1).expand(
+                rois.size(0), rois.size(1), 1)
+        rois = torch.cat([batch_index, rois[..., :4]], dim=-1)
+        batch_size = rois.shape[0]
+        num_proposals_per_img = rois.shape[1]
+
+        # Eliminate batch
```
Comment (jshilong) [inline mmdet/models/roi_heads/test_mixins.py:87]: eliminate the batch dimension

File: mmdet/models/roi_heads/test_mixins.py [hunk 4]
```diff
@@ -56,50 +56,93 @@ def simple_test_bboxes(self,
...
+        bbox_results = self._bbox_forward(x, rois)
         cls_score = bbox_results['cls_score']
         bbox_pred = bbox_results['bbox_pred']
-        # use shape[] to keep tracing
-        num_proposals_per_img = tuple(p.shape[0] for p in proposals)
-        rois = rois.split(num_proposals_per_img, 0)
-        cls_score = cls_score.split(num_proposals_per_img, 0)
 
-        # some detector with_reg is False, bbox_pred will be None
+        # Recover batch
```
Comment (jshilong) [inline mmdet/models/roi_heads/test_mixins.py:93]: Recover the batch dimension

File: mmdet/models/roi_heads/test_mixins.py [hunk 5]
```diff
@@ -56,50 +56,93 @@ def simple_test_bboxes(self,
...
+        rois = rois.reshape(batch_size, num_proposals_per_img, -1)
+        cls_score = cls_score.reshape(batch_size, num_proposals_per_img, -1)
+
+        if not torch.onnx.is_in_onnx_export():
+            # remove padding
+            supplement_mask = rois[..., -1] == 0
+            cls_score[supplement_mask, :] = 0
+
+        # some detector with_reg is False like Grid R-CNN,
```
Comment (jshilong) [inline mmdet/models/roi_heads/test_mixins.py:102]: bbox_pred would be None in some detector when with_reg is False, e.g. Grid R-CNN.

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 6]
```diff
@@ -275,35 +275,95 @@ def get_bboxes(self,
                    scale_factor,
                    rescale=False,
                    cfg=None):
+        """Transform network output for a batch into bbox predictions.
+
+        Args:
+            rois (Tensor): Boxes to be transformed. Has shape (N, 4)
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:281]: This maybe wrong

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 7]
```diff
@@ -275,35 +275,95 @@ def get_bboxes(self,
...
+               or (B, N, 4)
+            cls_score (list[Tensor] or Tensor): Box scores for
+              each scale level, each is a 4D-tensor, the channel number is
+              num_points * num_classes.
+            bbox_pred (Tensor, optional): Box energies / deltas for each scale
+                level, each is a 4D-tensor, the channel number is
+                num_classes * 4.
+            img_shape (Sequence[int] or torch.Tensor or Sequence[
+                Sequence[int]],optional): Maximum bounds for boxes, specifies
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:290]: Sequence[int]], optional)

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 8]
```diff
@@ -275,35 +275,95 @@ def get_bboxes(self,
...
+            det_bboxes = []
+            det_labels = []
+            for (bbox, score) in zip(bboxes, scores):
+                det_bbox, det_label = multiclass_nms(bbox, score,
+                                                     cfg.score_thr, cfg.nms,
+                                                     cfg.max_per_img)
+                det_bboxes.append(det_bbox)
+                det_labels.append(det_label)
+
+        if not is_batch:
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:363]: check this

File: mmdet/models/roi_heads/test_mixins.py [hunk 9]
```diff
@@ -55,51 +55,113 @@ def simple_test_bboxes(self,
...
+        Args:
+            x (tuple[Tensor]): Feature maps of all scale level.
+            img_metas (list[dict]): Image meta info.
+            proposals (Tensor or List[Tensor]): Region proposals.
+            rcnn_test_cfg (obj:`ConfigDict`): `test_cfg` of R-CNN.
+            rescale (bool): If True, return boxes in original image space.
+                Default: False.
+
+        Returns:
+            tuple[list[Tensor], list[Tensor]] or tuple[Tensor, Tensor] or
```
Comment (jshilong) [inline mmdet/models/roi_heads/test_mixins.py:69]: Does it support process different type of `Returns`?

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 10]
```diff
@@ -275,35 +275,95 @@ def get_bboxes(self,
...
-                bboxes[:, [1, 3]].clamp_(min=0, max=img_shape[0])
+                max_shape = bboxes.new_tensor(img_shape)[..., :2]
+                min_xy = bboxes.new_tensor(0)
+                max_xy = torch.cat(
+                    [max_shape] * 2, dim=-1).flip(-1).unsqueeze(-2)
+                bboxes = torch.where(bboxes < min_xy, min_xy, bboxes)
+                bboxes = torch.where(bboxes > max_xy, max_xy, bboxes)
+
+        if bboxes.ndim == 2:
+            is_batch = False
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:328]: comment why add this flag

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 11]
```diff
@@ -275,35 +275,97 @@ def get_bboxes(self,
...
+            cfg (obj:`ConfigDict`): `test_cfg` of Bbox Head. Default: None
+
+        Returns:
+            tuple[list[Tensor], list[Tensor]] or tuple[Tensor, Tensor]: The
+                first list includes batch tensor and the shape is (n, 5)
+                where 5 represent (tl_x, tl_y, br_x, br_y, score) and the
+                score between 0 and 1. The second list also includes tensors
+                of the same length as the first list, and the shape is (n,)
+                where each element represents the class label of the
+                corresponding box.
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:307]: The first list contains the boxes of all images in a  batch，each tensor has the shape  
 (num_bboexes, 5) and last dimension 5 represent (tl_x, tl_y, br_x, br_y, score). Each Tensor in the second list is the labels with shape (num_bboexes, ).

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 12]
```diff
@@ -275,35 +275,101 @@ def get_bboxes(self,
...
+               or (B, num_boxes, 5)
+            cls_score (list[Tensor] or Tensor): Box scores for
+               each scale level, each is a 4D-tensor, the channel number is
+               num_points * num_classes.
+            bbox_pred (Tensor, optional): Box energies / deltas for each scale
+                level, each is a 4D-tensor, the channel number is
+                num_classes * 4.
+            img_shape (Sequence[int] or torch.Tensor or Sequence[
+                Sequence[int]], optional): Maximum bounds for boxes, specifies
+                (H, W, C) or (H, W). If rois shape is (B, N, 4), then
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:295]: If rois shape is (B, num_boxes, 4)

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 13]
```diff
@@ -275,35 +275,101 @@ def get_bboxes(self,
...
+                the max_shape should be a Sequence[Sequence[int]]
+                and the length of max_shape should also be B.
+            scale_factor (ndarray): Scale factor of the image arange as
+                (w_scale, h_scale, w_scale, h_scale).
+            rescale (bool): If True, return boxes in original image space.
+                Default: False.
+            cfg (obj:`ConfigDict`): `test_cfg` of Bbox Head. Default: None
+
+        Returns:
+            tuple[list[Tensor], list[Tensor]] or tuple[Tensor, Tensor]: The
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:305]: If the input has a batch dimension, the return value is a tuple of the list. The first element is a list, which contains the boxes of the corresponding image in a batch, each tensor has the shape (num_boxes, 5) and last dimension 5 represent (tl_x, tl_y, br_x, br_y, score). Each Tensor in the second list is the labels with shape (num_boxes, ). The length of both lists should be equal to batch_size. Otherwise return value is a tuple of two tensors, first tensor is the boxes with scores, the second tensor is the labels, both have a shape with the first case.

File: tests/test_models/test_roi_heads/test_bbox_head.py [hunk 14]
```diff
@@ -64,6 +65,32 @@ def test_bbox_head_loss():
     assert losses.get('loss_bbox', 0) > 0, 'box-loss should be non-zero'
 
 
+@pytest.mark.parametrize(['num_sample', 'num_batch'], [[2, 2], [2, 0], [2, 0]])
```
Comment (jshilong) [inline tests/test_models/test_roi_heads/test_bbox_head.py:68]: add a num_sample is 0
and num_sample, num_batch both are 0.

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 15]
```diff
@@ -275,35 +275,100 @@ def get_bboxes(self,
...
+                bbox_pred = bbox_pred.unsqueeze(0)
+            rois = rois.unsqueeze(0)
 
         if bbox_pred is not None:
             bboxes = self.bbox_coder.decode(
-                rois[:, 1:], bbox_pred, max_shape=img_shape)
+                rois[..., 1:], bbox_pred, max_shape=img_shape)
         else:
-            bboxes = rois[:, 1:].clone()
+            bboxes = rois[..., 1:].clone()
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:339]: Why `clone`?

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 16]
```diff
@@ -275,35 +275,100 @@ def get_bboxes(self,
...
+                min_xy = bboxes.new_tensor(0)
+                max_xy = torch.cat(
+                    [max_shape] * 2, dim=-1).flip(-1).unsqueeze(-2)
+                bboxes = torch.where(bboxes < min_xy, min_xy, bboxes)
+                bboxes = torch.where(bboxes > max_xy, max_xy, bboxes)
+
+        if rescale and bboxes.size(-2) > 0:
+            if not isinstance(scale_factor, tuple):
+                scale_factor = tuple([scale_factor])
+            scale_factor = bboxes.new_tensor(scale_factor).unsqueeze(1).repeat(
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:351]: comment the shape here

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 17]
```diff
@@ -275,35 +275,100 @@ def get_bboxes(self,
...
-                bboxes /= scale_factor
+                max_shape = bboxes.new_tensor(img_shape)[..., :2]
+                min_xy = bboxes.new_tensor(0)
+                max_xy = torch.cat(
+                    [max_shape] * 2, dim=-1).flip(-1).unsqueeze(-2)
+                bboxes = torch.where(bboxes < min_xy, min_xy, bboxes)
+                bboxes = torch.where(bboxes > max_xy, max_xy, bboxes)
+
+        if rescale and bboxes.size(-2) > 0:
+            if not isinstance(scale_factor, tuple):
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:349]: update the docs if it is tuple by default.

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py (hunk 15 above)
Comment (hhaAndroid) [reply mmdet/models/roi_heads/bbox_heads/bbox_head.py:339]: I don’t know, it seems unnecessary.

**Review by jshilong — APPROVED** [review]
LGTM
